source/client/service_impl.cc

```diff
@@ -69,8 +69,12 @@ ::grpc::Status ServiceImpl::ExecutionStream(
   while (stream->Read(&request)) {
     ENVOY_LOG(debug, "Read ExecutionRequest data: {}", request.DebugString());
     if (request.has_start_request()) {
+      // It is possible to receive a back-to-back request here, while the future that is associated
+      // to our previous response is still active. To resolve this in practice we allow waiting here
+      // for a short period in favor of adding a more complex atomic bool to track if we are ready
+      // for the next request.
       if (future_.valid() &&
-          future_.wait_for(std::chrono::seconds(0)) != std::future_status::ready) {
+          future_.wait_for(std::chrono::seconds(1)) != std::future_status::ready) {
```

Comment:
Is there any way to do proper synchronization here that is race free? This seems an improvement, but our build/test machines sometimes encounter heavy load and we can flake with this.

Reply:
Yes there is, I just wanted to test the waters for this approach because it's the simplest one. I'll update it to eliminate the race altogether.